File: src/main/java/dev/ambryn/task/interfaces/State.java
package dev.ambryn.task.interfaces;

import com.fasterxml.jackson.annotation.JsonIgnore;
import dev.ambryn.task.models.Status;
import dev.ambryn.task.models.Task;

public abstract class State {
    @JsonIgnore
    protected Task context;
    protected Status status;

    public void setContext(Task context) {
        this.context = context;
    }

    public Status getStatus() {
        return this.status;
    }

    public abstract void start();

    public abstract void error();

    public abstract void finish();

    public abstract void cancel();
}

File: src/main/java/dev/ambryn/task/interfaces/State.kt
package dev.ambryn.task.interfaces

import dev.ambryn.task.models.Status
import dev.ambryn.task.models.Task

abstract class State(protected val context: Task) {
    abstract var status: Status
        protected set

    open fun start(): Unit {}

    open fun error(): Unit {}
    open fun finish(): Unit {}
    open fun cancel(): Unit {}
}
